rclcpp_action/src/client.cpp

```diff
@@ -12,30 +12,350 @@
+  const int64_t & sequence_number = response_header.sequence_number;
+  if (pimpl_->pending_result_responses.count(sequence_number) == 0) {
+    RCLCPP_ERROR(pimpl_->logger, "unknown result response, ignoring...");
+    return;
+  }
+  pimpl_->pending_result_responses[sequence_number](response);
+  pimpl_->pending_result_responses.erase(sequence_number);
+}
+
+void
+ClientBase::send_result_request(std::shared_ptr<void> request, ResponseCallback callback)
+{
+  std::lock_guard<std::mutex> guard(pimpl_->result_requests_mutex);
+  int64_t sequence_number;
+  rcl_ret_t ret = rcl_action_send_result_request(
+    pimpl_->client_handle.get(), request.get(), &sequence_number);
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret, "failed to send result request");
+  }
+  assert(pimpl_->pending_result_responses.count(sequence_number) == 0);
+  pimpl_->pending_result_responses[sequence_number] = callback;
+}
+
+void
+ClientBase::handle_cancel_response(
+  const rmw_request_id_t & response_header,
+  std::shared_ptr<void> response)
+{
+  std::lock_guard<std::mutex> guard(pimpl_->cancel_requests_mutex);
+  const int64_t & sequence_number = response_header.sequence_number;
+  if (pimpl_->pending_cancel_responses.count(sequence_number) == 0) {
+    RCLCPP_ERROR(pimpl_->logger, "unknown cancel response, ignoring...");
+    return;
+  }
+  pimpl_->pending_cancel_responses[sequence_number](response);
+  pimpl_->pending_cancel_responses.erase(sequence_number);
+}
+
+void
+ClientBase::send_cancel_request(std::shared_ptr<void> request, ResponseCallback callback)
+{
+  std::lock_guard<std::mutex> guard(pimpl_->cancel_requests_mutex);
+  int64_t sequence_number;
+  rcl_ret_t ret = rcl_action_send_cancel_request(
+    pimpl_->client_handle.get(), request.get(), &sequence_number);
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret, "failed to send cancel request");
+  }
+  assert(pimpl_->pending_cancel_responses.count(sequence_number) == 0);
+  pimpl_->pending_cancel_responses[sequence_number] = callback;
+}
+
+GoalID
+ClientBase::generate_goal_id()
+{
+  GoalID goal_id;
+  // TODO(hidmic): Do something better than this for UUID generation.
+  // std::generate(
+  //   goal_id.uuid.begin(), goal_id.uuid.end(),
+  //   std::ref(pimpl_->random_bytes_generator));
+  std::generate(
+    goal_id.begin(), goal_id.end(),
+    std::ref(pimpl_->random_bytes_generator));
+  return goal_id;
+}
+
+void
+ClientBase::execute()
+{
+  if (pimpl_->is_feedback_ready) {
+    std::shared_ptr<void> feedback_message = this->create_feedback_message();
+    rcl_ret_t ret = rcl_action_take_feedback(
+      pimpl_->client_handle.get(), feedback_message.get());
+    pimpl_->is_feedback_ready = false;
+    if (RCL_RET_OK != ret) {
+      rclcpp::exceptions::throw_from_rcl_error(ret, "error taking feedback");
+    } else {
+      this->handle_feedback_message(feedback_message);
+    }
+  }
+
+  if (pimpl_->is_status_ready) {
```

Comment:
The action server impl is only processing one item per `execute()` invocation ([ref](https://github.com/ros2/rclcpp/pull/593/files#diff-d222b3f55238d27bfbe118bd4125e4b2R171)). I'm not sure what advantage there is either way, but I think we should be consistent.

Reply:
Advantage of doing multiple things per execute: Don't have the overhead of calling `rcl_wait` again before the next thing is processed.

Advantage of doing one thing per execute: A multithreaded executor can do each thing in parallel.

For the server where someone is likely to try to block (though they shouldn't) in the `handle_accepted` callback, doing only one thing per `execute` makes it possible to use blocking callbacks with a multithreaded executor and a reentrant callback group. It might not be as important in the client, so I'm fine either way.

Reply:
One at a time execution in 8b173d5d2a3553318ca747c075cba250452a4a61